pkg/cloud/services/ec2/subnets.go

```diff
@@ -54,80 +56,66 @@ func (s *Service) reconcileSubnets() error {
 		return err
 	}
 
-	// If the subnets are empty, populate the slice with the default configuration.
-	// Adds a single private and public subnet in the first available zone.
-	if len(existing) < 2 && len(subnets) < 2 {
-		zones, err := s.getAvailableZones()
-		if err != nil {
-			return err
-		}
-
-		if len(subnets.FilterPrivate()) == 0 {
-			if s.scope.VPC().IsUnmanaged(s.scope.Name()) {
-				return errors.New("expected at least one private subnet available for use, got 0")
-			}
+	unmanagedVPC := s.scope.VPC().IsUnmanaged(s.scope.Name())
 
-			subnets = append(subnets, &infrav1.SubnetSpec{
-				CidrBlock:        defaultPrivateSubnetCidr,
-				AvailabilityZone: zones[0],
-				IsPublic:         false,
-			})
+	if len(subnets) == 0 {
```

Comment:
What happens in the case that we hit a transient error during initial reconciliation before we are able to create all of the subnets returned by `s.getDefaultSubnets()`? Would we only detect any subnets that were already created and persisted into the AWSCluster.Spec from the first reconciliation?

What about if no subnets get persisted back to the AWSClusterSpec, do we need to worry about potential subnet conflicts when trying to reconcile a second time?

Or if there is an error during initial reconciliation and no subnets are persisted back to the AWSClusterSpec for some reason (say a controller crash or deployment update happens at an inopportune time) could we possibly end up with multiple subnets being attached to the same AZ (when `AZSelectionSchemeRandom` is selected)?

Reply:
Thats a good point @detiber. 

Short term the easiest way to guard against this is to exit from the subnet/network reconcilation look after `subnets, err = s.getDefaultSubnets()` 
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/services/ec2/subnets.go#L70
so that the subnets are are saved into the NetworkSpec. Then on the next reconciliation loop it will create the subnets in the spec (like if you have an unmanaged VPC).

Longer term perhaps it would be better to move the `s.getDefaultSubnets()` into the defaulting webhook.

Shall i do the small shorter term change and then create a issue for moving the defaulting into the webhook?

Reply:
+1 to the change to returning early after the call to getDefaultSubnets and relying on a subsequent reconciliation to create the subnets.

I do like the idea of handling it through a webhook, however I don't think we'd be able to do that with the current multi-tenancy model, since the webhook is running without ec2 credentials. Perhaps that is something that we could do after we support native multi-tenancy. @randomvariable do you have any thoughts about that?